ffi/src/handle.rs

```diff
@@ -0,0 +1,174 @@
+//! Logic related to opaque pointer handles and functions that are common to multiple handle types
+
+/// Type of the handle id
+pub type Id = u8;
+
+/// Implement [`$crate::Handle`] for given types with the given initial handle id.
+#[macro_export]
+macro_rules! handles {
```

Comment:
I wonder why we need handles?
Why not have methods like A__clone, B__clone, ...?
Having an IDs seems a bit error prone.

Reply:
I mean that that for example for inherit methods we generate methods like `Type__method_name`, could that be the case for `Clone`, `Drop`, ...
So insead of having handle ids we would have separate functions for each type.

Reply:
Yes, that's the more safe alternative. And you might be correct it's a better approach. However, it introduces function declaration bloat in the binary/smartcontract so I went with multiplexing approach. I can turn this into an issue to discuss further and change the implementation

Reply:
Handle ids are essentially a method multiplexing mechanism

Reply:
Okay, I see it makes sense to reduce the memory footprint of the smart contract.

Thank you for the clarification.

> I can turn this into an issue to discuss further and change the implementation 

If you think this is necessary. 

Last question: 
Types in different crates could have overlapping handle ids because underling `__clone` is different for different crates, right?